rfcs/20200721-extension-types.md

```diff
@@ -0,0 +1,1304 @@
+
+Job                         | SimpleMaskedTensorSpec
+--------------------------- | -----------------------------------------------
+Storing non-tensor metadata | Stores the shape and value dtype for the masked tensor.
+Serializing the TypeSpec    | Serializes the shape and dtype as a tuple.
+Decomposing values          | Decomposes the masked tensor into a (value, mask) tuple.
+Checking compatibility      | Considers two MaskedTensors compatible if their dtypes and shapes are compatible.
+
+The complete code for `SimpleMaskedTensorSpec` is shown below:
+
+```python
+class SimpleMaskedTensorSpec(tf.TypeSpec):
+  """Type specification for a `SimpleMaskedTensor`."""
+
+  def __init__(self, shape: tf.TensorShape, dtype: tf.dType):
+    """Creates a new `SimpleMaskedTensorSpec`.
+
+    Args:
+      shape: The shape of the `SimpleMaskedTensor`.
+      dtype: The dtype of the `SimpleMaskedTensor`'s values.
+    """
+    self._shape = shape
+    self._dtype = dtype
+
+  shape = property(lambda self: self._shape)
+  dtype = property(lambda self: self._dtype)
+  value_type = property(lambda: SimpleMaskedTensor)
+
+  def to_components(self, masked_tensor):
+    return (masked_tensor.value, masked_tensor.mask)
+
+  def from_components(self, components):
+    return SimpleMaskedTensor(*components)
+
+  def component_specs(self):
+    return (tf.TensorSpec(self._shape, self._dtype),
+            tf.TensorSpec(self._shape, tf.bool))
+
+  def serialize(self):
+    return (self._shape, self._dtype)
+```
+
+**Note:** `SimpleMaskedTensorSpec` uses the default implementations for several
+`TypeSpec` methods, such as `is_compatible`, which are defined based on
+`serialize` and `deserialize`.
+
+At this point, `SimpleMaskedTensor` can be used with "generic" TensorFlow APIs,
+such as `tf.function`, `SavedModel`, and `tf.while_loop`.  But since
+`SimpleMaskedTensor` is tensor-like, it makes sense for it to implement the
+`tf.DispatchableType` protocol as well.  We can do so by adding a
+`__tf_dispatch__` method.  For simplicity, we will only show support for unary
+and binary elementwise operations and a handful of other operations in this
+example.
+
+
+```python
+class SimpleMaskedTensor(object):
+  # [...definition continued from above...]
+
+  @classmethod
+  def __tf_dispatch__(cls, op, args, kwargs):
+    if tf.dispatch.is_unary_elementwise_op(op):
+      return self._unary_elementwise_dispatch(op, args, kwargs)
+    elif tf.dispatch.is_binary_elementwise_op(op):
+      return self._binary_elementwise_dispatch_op, args, kwargs)
+    else:
+      dispatch_handler = SimpleMaskedTensor.__dispatchers.get(op, None)
+      if dispatch_hander is not None:
+        return dispatch_handler(args, kwargs)
+      else:
+        return NotImplemented
+
+  # We support ops that take tf.Tensor and SimpleMaskedTensor arguments.  We
+  # don't support any other dispatchable argument types (such as tf.RaggedTensor).
+  __tf_dispatch_types__ = (tf.Tensor, SimpleMaskedTensor)
```

Comment:
What about builtin numeric types, like `int` or `float`. Would these need to be included explicitly in `__tf_dispatch_types__`?

Reply:
In the current design, arguments that don't define `__tf_dispatch__` will be converted to `Tensor` before dispatch is called.  So ints, floats, np.arrays, etc., will get converted to `Tensor`.